### ISSUE TITLE
docker ps hang and not able to create containers

### ISSUE THREAD

**Issue by nkkashyap** [moby/moby#27543]
**Description**

Docker1.12.0 is hanging and not able to create container (docker run) or list the containers (docker list). Stack dump shows one of the go routine stuck in device-mapper call

```
goroutine 226734806 [syscall, 67 minutes, locked to thread]:
github.com/docker/docker/pkg/devicemapper._Cfunc_dm_udev_wait(0xca0d4d2d0c, 0xca00000000)
    ??:0 +0x41
github.com/docker/docker/pkg/devicemapper.dmUdevWaitFct(0xd4d2d0c, 0x441a8e)
    /go/src/github.com/docker/docker/pkg/devicemapper/devmapper_wrapper.go:222 +0x22
github.com/docker/docker/pkg/devicemapper.UdevWait(0xca167ab4f0, 0x0, 0x0)
    /go/src/github.com/docker/docker/pkg/devicemapper/devmapper.go:259 +0x3b
github.com/docker/docker/pkg/devicemapper.activateDevice(0xcaa2eda990, 0x22, 0xcc57547380, 0x57, 0xbdf1, 0x280000000, 0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/docker/docker/pkg/devicemapper/devmapper.go:750 +0x8b8
github.com/docker/docker/pkg/devicemapper.ActivateDevice(0xcaa2eda990, 0x22, 0xcc57547380, 0x57, 0xbdf1, 0x280000000, 0x0, 0x0)
    /go/src/github.com/docker/docker/pkg/devicemapper/devmapper.go:711 +0x78
github.com/docker/docker/daemon/graphdriver/devmapper.(*DeviceSet).activateDeviceIfNeeded(0xc82009a9c0, 0xc8fc801080, 0xc82009a900, 0x0, 0x0)
    /go/src/github.com/docker/docker/daemon/graphdriver/devmapper/deviceset.go:539 +0x58d
github.com/docker/docker/daemon/graphdriver/devmapper.(*DeviceSet).MountDevice(0xc82009a9c0, 0xc85abb2410, 0x45, 0xc919beee00, 0x76, 0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/docker/docker/daemon/graphdriver/devmapper/deviceset.go:2269 +0x2cd
github.com/docker/docker/daemon/graphdriver/devmapper.(*Driver).Get(0xc82097abe0, 0xc85abb2410, 0x45, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/docker/docker/daemon/graphdriver/devmapper/driver.go:185 +0x62f
github.com/docker/docker/daemon/graphdriver.(*NaiveDiffDriver).Get(0xc820882700, 0xc85abb2410, 0x45, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    <autogenerated>:30 +0xa6
github.com/docker/docker/layer.(*layerStore).initMount(0xc820882880, 0xcd94d93c00, 0x40, 0xc8201e4c40, 0x40, 0x0, 0x0, 0xcb75851c10, 0xcac667ec90, 0x0, ...)
    /go/src/github.com/docker/docker/layer/layer_store.go:589 +0x274
github.com/docker/docker/layer.(*layerStore).CreateRWLayer(0xc820882880, 0xcd94d92fc0, 0x40, 0xc85abb23c0, 0x47, 0x0, 0x0, 0xcb75851c10, 0xcac667ec90, 0x0, ...)
    /go/src/github.com/docker/docker/layer/layer_store.go:469 +0x46a
github.com/docker/docker/daemon.(*Daemon).setRWLayer(0xc82009a340, 0xc9ae5bc960, 0x0, 0x0)
    /go/src/github.com/docker/docker/daemon/create.go:207 +0x203
github.com/docker/docker/daemon.(*Daemon).create(0xc82009a340, 0xcc381ff223, 0x29, 0xca15f888c0, 0xc84f4ac380, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/docker/docker/daemon/create.go:106 +0x3ec
github.com/docker/docker/daemon.(*Daemon).containerCreate(0xc82009a340, 0xcc381ff223, 0x29, 0xca15f888c0, 0xc84f4ac380, 0x0, 0xca2f975500, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/docker/docker/daemon/create.go:55 +0x4c3
github.com/docker/docker/daemon.(*Daemon).ContainerCreate(0xc82009a340, 0xcc381ff223, 0x29, 0xca15f888c0, 0xc84f4ac380, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/docker/docker/daemon/create.go:29 +0xc2
```

**Steps to reproduce the issue:**
Issue is occurring randomly and occasionally

**Output of `docker version`:**

```
docker version
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   535b3ac-unsupported
 Built:        Tue Sep  6 10:25:57 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   535b3ac-unsupported
 Built:        Tue Sep  6 10:25:57 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Server Version: 1.12.0

Storage Driver: devicemapper
 Pool Name: docker-8:6-1572870-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/vg-docker/data
 Metadata file: /dev/vg-docker/metadata
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true

Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 4.4.0-34-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

```
lvm2                             2.02.98-6ubuntu2
libdevmapper1.02.1:amd64         2:1.02.90-2ubuntu1  
```

**Comment by cpuguy83** [comment]
ping @rhvgoyal

**Comment by rhvgoyal** [comment]
This activation of device seems to be waiting on UdevWait(). That is waiting on a cookie from libdevmapper point of view. Which is basically a semaphore on which a put() operation is done
by udev rules when certain event happens. So in this case, somehow event has not happened, or
rule has not executed or something else went wrong, so library is waiting infinitely, hence docker is waiting infinitely, hence blocking everything else behind it.

Now question is, why did we reach in this stage, and there are no good answers. So many thing could be wrong. There are many components which could have some kind of issue.

Are you able to reproduce it?

`dmsetup udevcookies` can show you the cookies libdm is waiting on. To force system out of this manually, you can try `dmsetup udevcomplete <cookie>`. This is artificial completion which actually should have happpened from udev rules.

Make sure you are running latest packages. Try rhel/fedora/centos and see if problem is reproducible therere too.

**Comment by rhvgoyal** [comment]
Check your logs. We have had cases where due to memory pressure, udev worker was terminated by OOM. That means a certain udev rule never executed, that means completion never happened on cookie and we experieced this hang. 

In your logs if you find that some udev work was killed, that might be contributing to this issue.

**Comment by nkkashyap** [comment]
@rhvgoyal Thanks Vivek
Scanned the kernel logs, no OOM message.
We are using Ubuntu 14.04, currently cannot move to other destro. Will try Docker 1.12.3 with lvm2_2.02.133

**Comment by depay** [comment]
@rhvgoyal completing cookie works for me. thanks !

**Comment by cpuguy83** [comment]
Another report for this on Ubuntu 14.04: https://github.com/docker/docker/issues/25993#issuecomment-288513872

I'd recommend upgrading to 14.04.5 which uses kernel 4.4 instead of 3.13, in which case you can just use overlay2 instead, but also more likely to have fixes for issues like this.

@nkkashyap Any news?

**Comment by thaJeztah** [comment]
closing as this one went stale
